services/system/StagedTx/plugin/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+            }}
+        }}"#,
+        id = id
+    );
+
+    let details = Server::post_graphql_get_json(&query).unwrap();
+    let details = serde_json::from_str::<StagedTxDetails>(&details).unwrap();
+    let txid = details.data.details.txid;
+    if txid.len() % 2 != 0 {
+        return Err(ErrorType::InvalidTxId.into());
+    }
+    let txid: Vec<u8> = (0..txid.len())
+        .step_by(2)
+        .map(|i| u8::from_str_radix(&txid[i..i + 2], 16).unwrap())
+        .collect();
+    let txid: [u8; 32] = txid.try_into().map_err(|_| ErrorType::InvalidTxId)?;
+
+    Ok(txid.into())
+}
+
+struct StagedTxPlugin;
+
+impl Proposer for StagedTxPlugin {
+    fn set_propose_latch(account: String) -> Result<(), Error> {
+        validate_account(&account)?;
+        hook_tx_transform_label(Some(account.as_str()));
+        Ok(())
+    }
+
+    fn remove(id: u32) -> Result<(), Error> {
+        add_action_to_transaction(
+            remove::ACTION_NAME,
+            &remove {
+                id,
+                txid: get_staged_txid(id)?,
+            }
+            .packed(),
+        )
+    }
+}
+
+impl Respondent for StagedTxPlugin {
+    fn accept(id: u32) -> Result<(), Error> {
+        add_action_to_transaction(
+            accept::ACTION_NAME,
+            &accept {
+                id,
+                txid: get_staged_txid(id)?,
+            }
+            .packed(),
+        )
+    }
+
+    fn reject(id: u32) -> Result<(), Error> {
+        add_action_to_transaction(
+            reject::ACTION_NAME,
+            &reject {
+                id,
+                txid: get_staged_txid(id)?,
+            }
+            .packed(),
+        )
+    }
+}
+
+impl HookTxTransform for StagedTxPlugin {
+    fn on_tx_transform(
+        label: Option<String>,
+        actions: Vec<Action>,
+    ) -> Result<Option<Vec<Action>>, Error> {
+        let Some(label) = label else {
+            return Ok(None);
+        };
+
+        let transact = psibase::services::transact::SERVICE.to_string();
+        get_assert_caller("on_tx_transform", &[&transact])?;
+
+        // There is typically only one sender for a set of actions, but in rare cases where
+        // hooks were used to simultaneously execute actions from multiple users *and* a propose
+        // latch was set, the result is:
+        // 1. Multiple staged transactions are proposed, one for each sender.
+        // 2. The actions contained in a staged transaction correspond to the actions originally executed by sender.
+        // 3. The new sender of the staged actions correspond to the latch that was set at the time the action was scheduled.
+
+        let action_groups = group_by_sender(actions);
+        let actions = action_groups
+            .into_iter()
+            .map(|group| propose_wrap(group, &label))
+            .collect();
+
+        Ok(Some(actions))
+    }
+}
+
+fn propose_wrap(mut actions: Vec<Action>, label: &str) -> Action {
+    let sender = actions[0].sender.clone();
```

Comment:
The logic here is pretty convoluted. I don't like the way the actions are created with the wrong sender and then patched up later. Note that having the wrong sender does not just mean that an action will fail. It could make the action mean something else entirely. Could we provide a way to set the sender explicitly when adding an action to the transaction, and have this only wrap the actions in `propose`?

Reply:
I decided to go for it. Can always roll it back, but I don't think it's actually a real security problem. 
Now plugins can ask `get-current-user`. I changed it from `get-logged-in-user` to avoid confusion over which user is being returned, the one logged in to "my plugin's app" or the one logged in to the top-level app. I think "current user" is clearer.